acapy_agent/core/plugin_registry.py

```diff
@@ -216,134 +253,183 @@ async def load_protocol_version(
+                LOGGER.trace("Attempting to load definition from: %s", definition_path)
+                definition = ClassLoader.load_module(definition_path)
             except ModuleLoadError as e:
                 LOGGER.error("Error loading plugin definition module: %s", e)
                 return
 
             if definition:
+                LOGGER.trace("Loading versioned protocols for: %s", plugin_name)
                 for protocol_version in definition.versions:
+                    version_path = (
+                        f"{plugin_name}.{protocol_version['path']}.message_types"
+                    )
                     try:
-                        mod = ClassLoader.load_module(
-                            f"{plugin.__name__}.{protocol_version['path']}"
-                            + ".message_types"
-                        )
-                        await self.load_protocol_version(context, mod, protocol_version)
-
+                        LOGGER.trace("Loading message types from: %s", version_path)
+                        mod = ClassLoader.load_module(version_path)
                     except ModuleLoadError as e:
-                        LOGGER.error("Error loading plugin module message types: %s", e)
+                        LOGGER.error(
+                            "Error loading plugin module message types from %s: %s",
+                            version_path,
+                            e,
+                        )
                         return
 
-    async def register_admin_routes(self, app):
+                    if mod:
+                        await self.load_protocol_version(context, mod, protocol_version)
+                    else:
+                        LOGGER.debug("Failed to load %s", version_path)
+
+    async def register_admin_routes(self, app) -> None:
         """Call route registration methods on the current context."""
+        LOGGER.trace("Registering admin routes for %d plugins", len(self._plugins))
+
         for plugin in self._plugins.values():
-            definition = ClassLoader.load_module("definition", plugin.__name__)
+            plugin_name = plugin.__name__
+            LOGGER.trace("Processing routes for plugin: %s", plugin_name)
+            mod = None
+            definition = ClassLoader.load_module("definition", plugin_name)
             if definition:
                 # Load plugin routes that are in a versioned package.
+                LOGGER.trace("Loading versioned routes for: %s", plugin_name)
                 for plugin_version in definition.versions:
+                    version_path = f"{plugin_name}.{plugin_version['path']}.routes"
                     try:
-                        mod = ClassLoader.load_module(
-                            f"{plugin.__name__}.{plugin_version['path']}.routes"
-                        )
+                        LOGGER.trace("Loading routes from: %s", version_path)
+                        mod = ClassLoader.load_module(version_path)
```

Comment:
We're overwriting `mod` with each version. Only the last mod's routes are actually registered.

Reply:
Aah, that's what I get for micro-optimizing! I moved the `await mod.register(app)` step out of the for loop without noticing ... because it seemed  duplicated in both if and else blocks... Thanks for picking up on that. 

These methods also get flagged by sonarcloud for their cognitive complexity, due to all the nesting. So I want to refactor that for readability too :+1: